### PR TITLE
Add timeout support on send

### PR DESCRIPTION
**Pull request by egorse**
Add internal/errorgrp package to support cancellable error groups
Add tests for push/pull timeout

### CODE REVIEW

**Review by sbinet — COMMENTED** [review]
thanks for putting this together.

according to codecov:
https://app.codecov.io/gh/go-zeromq/zmq4/pull/148/blob/internal/errorgrp/errorgrp.go

we don't exercize the whole `errorgrp.Group2.Go` code.
we could probably try to make sure we still get the context error when it expires *during* the execution of `f`.
(probably with a 2-level handshake where the `cancel()` happens when we are sure `f` is being executed)

**Comment by egorse** [comment]
> we don't exercize the whole `errorgrp.Group2.Go` code.
> we could probably try to make sure we still get the context error when it expires _during_ the execution of `f`.
> (probably with a 2-level handshake where the `cancel()` happens when we are sure `f` is being executed)

This is sort of odd - I had impression that only block execution of f if context is already cancelled (i.e. if the eg has limits on number of executable functions and parent context got cancelled prior next f in eg started).
I will check it and update PR (include others suggestions)

File: internal/errorgrp/errorgrp.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+// Package errorgrp is bit more advanced than errgroup
+// Major difference is that when error group is created with WithContext2
+// the parent context would implicitly cancel all functions called by Go method.
+//
+// The name is selected so you can mix regular errgroup and errorgrp in same file.
+package errorgrp
```
[the remark this answers is not in the document]
Comment (egorse) [reply internal/errorgrp/errorgrp.go:6]: Commonly used ```g := new(errgroup.Group)``` became tricky to handle in new internal/errgroup. I will try to make

**Comment by egorse** [comment]
I think its ready for next review cycle. Please comment

File: internal/errgroup/errgroup.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+type Group struct {
+	grp *errgroup.Group
+	ctx context.Context
+}
+
+// WithContext creates Group and store inside parent context
+// so the Go method would respect parent context cancellation
+func WithContext(ctx context.Context) (*Group, context.Context) {
+	grp, child_ctx := errgroup.WithContext(ctx)
+	return &Group{grp: grp, ctx: ctx}, child_ctx
```
[the remark this answers is not in the document]
Comment (egorse) [reply internal/errgroup/errgroup.go:27]: We are storing parent context as have to check its cancellation - not the child ctx. The child context would be cancelled when the errgroup is done.

The copy by value is very nice to get rid of init

File: internal/errgroup/errgroup.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+// The init method provides backward compatibility to x/sync/errgroup.Group
+// when developers can create error group by i.e. ```eg := &errgroup.Group{}```
+func (g *Group) init() {
+	if g.grp == nil {
+		g.grp = &errgroup.Group{}
+	}
+}
+
+func (g *Group) wrap(f func() error) func() error {
+	return func() error {
```
[the remark this answers is not in the document]
Comment (egorse) [reply internal/errgroup/errgroup.go:86]: I will correct it today\tomorrow

**Comment by egorse** [comment]
Please take a look. Still not sure how well it is on implicit errgroup.Group initialization

**Comment by egorse** [comment]
Dough - original init() was nicer for codecov. Will wait for your comments

**Review by sbinet — APPROVED** [review]
LGTM

thanks a lot for the PR.

**Pull request merged**